apps/wing-console/console/ui/src/layout/default-layout.tsx

```diff
@@ -239,6 +245,43 @@ export const DefaultLayout = ({
     ],
   );
 
+  const [currentNodeMetadata, setCurrentNodeMetadata] = useState<
+    string | undefined
+  >();
+  const [metadataInstances, setMetadataInstances] = useState<
+    Record<string, MetadataInstance>
+  >({});
+
+  useEffect(() => {
+    const currentNode = selectedItems[0];
+    if (!currentNode) {
+      return;
+    }
+
+    setMetadataInstances((instances) => {
+      instances[currentNode] = {
+        id: currentNode,
+        data: metadata.data,
+      };
+      return instances;
+    });
+    setCurrentNodeMetadata(currentNode);
+  }, [metadata.data, selectedItems]);
+
+  useEffect(() => {
+    if (!nodeIds.data) {
+      return;
+    }
+    setMetadataInstances((instances) => {
```

Comment:
I would add a comment describing this logic.
If I understand correctly, this cleans up all removed resources from the `metadataInstances` array, right?